### PR TITLE
Convert index.js and team-validator-async.js to typescript

### PR DESCRIPTION
**Pull request by not-a-seagull**

### CODE REVIEW

File: server/index.ts [hunk 1]
```diff
@@ -35,13 +35,10 @@
  * Sockets - from sockets.js
  *
  *   Used to abstract out network connections. sockets.js handles
- *   the actual server and connection set-up.
```
Comment (Zarel) [inline server/index.ts]: ? I assume this was a mistake

File: server/index.ts [hunk 2]
```diff
@@ -54,24 +51,29 @@ try {
...
 }
+
 try {
 	require.resolve('../.sim-dist/index');
 } catch (e) {
 	throw new Error("Dependencies are unmet; run `node build` before launching Pokemon Showdown again.");
 }
 
-const FS = require('../.lib-dist/fs').FS;
+import { FS } from '../lib/fs';
```
Comment (Zarel) [inline server/index.ts]:
```suggestion
import {FS} from '../lib/fs';
```
(style)

File: server/index.ts (hunk 1 above)
Comment (not-a-seagull) [reply server/index.ts]: Yeah, I don't know what that's doing there.

File: server/index.ts [hunk 3]
```diff
@@ -54,24 +52,29 @@ try {
...
-const FS = require('../.lib-dist/fs').FS;
+import {FS} from '../lib/fs';
 
 /*********************************************************
  * Load configuration
  *********************************************************/
 
-const ConfigLoader = require('../.server-dist/config-loader');
+// global becomes much easier to use if declared as an object
+declare var global: any;
```
Comment (Zarel) [inline server/index.ts]:
```suggestion
declare const global: any;
```

We shouldn't be changing where it points to, anyway.

File: server/team-validator-async.ts [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+
+export const get = TeamValidatorAsync.get;
+
+/*********************************************************
+ * Process manager
+ *********************************************************/
+
+import { QueryProcessManager } from '../lib/process-manager';
+
+// @ts-ignore
```
Comment (Zarel) [inline server/team-validator-async.ts]: Wait, what's being suppressed here?

File: server/team-validator-async.ts [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+export const PM = new QueryProcessManager(module, async message => {
```
Comment (Zarel) [inline server/team-validator-async.ts]: Why `async`?

Comment (Zarel) [reply server/team-validator-async.ts]: Looks like it was in the original code. I'm fine with leaving it alone.

File: server/team-validator-async.ts (hunk 4 above)
Comment (not-a-seagull) [reply server/team-validator-async.ts]: It was there in the original code. When I removed it, TSC didn't throw any errors.

File: server/index.ts [hunk 6]
```diff
@@ -54,24 +52,29 @@ try {
...
 
-const ConfigLoader = require('../.server-dist/config-loader');
+// global becomes much easier to use if declared as an object
+declare const global: any;
+
+import * as ConfigLoader from './config-loader';
 global.Config = ConfigLoader.Config;
 
-global.Monitor = require('../.server-dist/monitor').Monitor;
+import { Monitor } from './monitor';
```
Comment (Zarel) [inline server/index.ts]:
```suggestion
import {Monitor} from './monitor';
```

Could you grep `import { ` (with the space) and fix those, too?

Comment (not-a-seagull) [reply server/index.ts]: Resolved this in the latest commit

**Comment by TheJetOU** [comment]
You'll also need to edit the `pokemon-showdown` script to refer to `.server-dist` when starting up the server.

**Comment by Zarel** [comment]
This will also be awkward for the next time we restart the server.

Might as well update to Node 13, right? Right???

**Comment by Zarel** [comment]
Ugh, I can't merge this for today's restart because of conflicts. :(

**Comment by TheJetOU** [comment]
> Ugh, I can't merge this for today's restart because of conflicts. :(

Thankfully, maintainers can push to branches that are being PR'ed :).

**Comment by Zarel** [comment]
@TheJetOU Unless it's merging suggestions, because of a GitHub bug. Anyway, I don't have time for complicated fixes during restarts anyway.

**Comment by not-a-seagull** [comment]
I fixed the merge conflicts.

**Comment by Zarel** [comment]
@TheJetOU brings up a good point, this might not require a restart. The hotpatchable/unhotpatchable interface doesn't seem to be crossed much here.

**Pull request merged**